woocommerce/woocommerce-gateway-stripe
# Set Klarna's preferred locale

Fixes #3087 

## Changes proposed in this Pull Request:
For a more consistent experience for the shopper, we would like to render the Klarna payments page in the store locale.

Stripe limits Klarna localization to [the set of supported locales for the billing country](https://docs.stripe.com/payments/klarna/accept-a-payment?web-or-mobile=web&payments-ui-type=direct-api#supported-locales-and-currencies). For cases where the exact store locale is not supported, we will use the closest one. For example, if the store locale is "fr_FR" and the billing country is "Canada", we will set the preferred locale to "fr-CA" instead of "fr-FR".

If the store locale's language is not supported for the billing country, e.g. "fr" for "US", we do not pass a preferred locale and let Klarna decide on the default behavior.







## Testing instructions
1. Assumption: your Stripe test account's country is USA.
   - https://dashboard.stripe.com/test/settings/account
2. Enable Klarna as a payment method in WC Stripe.
3. Choose "Español" as your store locale: `wp-admin/options-general.php`
4. As a shopper, add a product and go to checkout.
5. Choose "United States" as your billing country.
6. Select Klarna as your payment option and click "Place Order".
7. **BEFORE** THIS PR: The Klarna payment page is in English.
8. **AFTER** THIS PR: The Klarna payment page is in Spanish.
9. To test Klarna's expected default behavior: set the store locale to "Français" and the billing country to "United States". The Klarna page should be in English.

**To test other combinations:**
- Choose a billing country for which Klarna supports at least two locales, e.g. Canada.
   - See [supported locales per billling country](https://docs.stripe.com/payments/klarna/accept-a-payment?web-or-mobile=web&payments-ui-type=direct-api#supported-locales-and-currencies) 
- Make sure your Stripe test account country is the same as your billing country, and the currency requirements are met, in order for Klarna to show up as an option during checkout.
- As your store locale, choose the locale that is not the default locale for the billing country. For example, for Canada, choose "Français du Canada" as your store locale. This is not a requirement; this is only so we can reasonably say that the behavior is because of our code, and not because of Klarna defaults.


## Screenshots

**With store locale as "es_ES", and billing country in "US", we set Klarna's preferred locale to "es-US"**
<img width="400" alt="Screenshot 2024-09-09 at 10 38 11 PM" src="https://github.com/user-attachments/assets/38990056-7742-44d2-976f-e1ae90eddb19">






---


**Post merge**
